config/prowgen/pkg/context.go

```diff
@@ -116,18 +113,75 @@ func (pc *ProwContext) Periodics(job *Job, periodicityHours int) {
 				BaseRef: pc.Branch,
 			},
 		},
-		Interval: strconv.Itoa(periodicityHours) + "h",
-		// TODO: use Cron instead of Interval
-		// Cron: pc.cronSchedule(periodicityHours),
+		PeriodicityHours: periodicityHours,
+		// Use a minute and startHour of 0 when adding the period here, but in JobFile
+		// when we actually generate the JobFile struct we'll recalculate values
+		// to spread the periodics evenly across every hour / day
+		Cron: cronSchedule(0, 0, periodicityHours),
 	})
 }
 
 func (pc *ProwContext) JobFile() *JobFile {
-	// TODO: when using Cron instead of Interval for periodics, adjust all periodics
-	// here to spread them evenly throughout the hour
-
 	presubmitKey := fmt.Sprintf("%s/%s", pc.Org, pc.Repo)
 
+	// By dividing 60 by the number of periodics we get the maximum number of minutes
+	// apart which we can schedule jobs, in an effort to maximise the amount of time
+	// that each job could theoretically run without anything else running in parallel.
+
+	// We aim to maximise the amount of CPU available to each job at startup, which is
+	// often when most jobs need a lot of CPU (e.g. to set up tests / clusters)
+	// We use ceil because more spreading isn't a bad thing
+	minuteSpread := int(math.Ceil(60.0 / float64(len(pc.periodics))))
+
+	// Count the number of jobs with each periodicity to make it easier to spread them later
+	periodicityCounts := map[int]int{}
+
+	for _, p := range pc.periodics {
+		periodicityCounts[p.PeriodicityHours] += 1
+	}
+
+	// periodicitySeen is used to track how many times we've seen a job with a given
+	// periodicity, which allows us to spread jobs with the same periodicity across the day
+	periodicitySeen := map[int]int{}
+
+	for i, p := range pc.periodics {
+		minute := (i * minuteSpread) % 60
```

Comment:
Currently, all branches will start at the same minute and hour, could we maybe add some kind of hash to "seed" the minute and hour calculation?

Reply:
Meant to add some context here as to what I did so I'll do that now:

I went with a counter of the number of files we'd generated. Using a hash to seed RNG or something like that runs the risk of creating collisions with ~4 branches we're processing and the small numbers we're using, so there's a decent chance we'd end up no better off than having every branch running at the same minute. With a counter, it should differ a bit more!